pkg/deviceplugins/vgpu_device_manager.go

```diff
@@ -0,0 +1,488 @@
+		}
+		monitoredDevices[dev.ID] = vgpuDevice
+	}
+
+	logrus.Infof("all monitored devices %s for plugin %s", monitoredDevices, dp.resourceName)
+	dirName = filepath.Dir(dp.socketPath)
+	err = watcher.Add(dirName)
+
+	if err != nil {
+		return fmt.Errorf("failed to add the device-plugin kubelet path to the watcher: %v", err)
+	}
+	_, err = os.Stat(dp.socketPath)
+	if err != nil {
+		return fmt.Errorf("failed to stat the device-plugin socket: %v", err)
+	}
+
+	err = watcher.Add(dp.socketPath)
+	if err != nil {
+		return fmt.Errorf("failed to watch device-plugin socket: %v", err)
+	}
+
+	return dp.performCheck(monitoredDevices, watcher)
+}
+
+// performCheck performs checks and monitors the devices
+func (dp *VGPUDevicePlugin) performCheck(monitoredDevices map[string]string, watcher *fsnotify.Watcher) error {
+	// run initial health check for devices created before. This works around device restarts
+	// the device plugin runs out of band from the actual device enablement so the first device could be missed by the plugin
+	for monDevID, devPath := range monitoredDevices {
+		_, err := os.Stat(devPath)
+		if err == nil {
+			logrus.Infof("marking devID %s healthy for plugin %s", monDevID, dp.resourceName)
+			dp.health <- deviceHealth{
+				DevID:  monDevID,
+				Health: pluginapi.Healthy,
+			}
+		}
+	}
+
+	for {
+		select {
+		case <-dp.stop:
+			return nil
+		case err := <-watcher.Errors:
+			logrus.Errorf("error watching devices and device plugin directory: %v", err)
+		case event := <-watcher.Events:
+			logrus.Infof("got event for device %s in plugin %s", event.Name, dp.resourceName)
+			if monDevID, exist := monitoredDevices[event.Name]; exist {
+				// Health in this case is if the device path actually exists
+				if event.Op == fsnotify.Create {
+					logrus.Debugf("monitored device %s appeared", dp.resourceName)
+					dp.health <- deviceHealth{
+						DevID:  monDevID,
+						Health: pluginapi.Healthy,
+					}
+				} else if (event.Op == fsnotify.Remove) || (event.Op == fsnotify.Rename) {
+					logrus.Debugf("monitored device %s disappeared", dp.resourceName)
+					dp.health <- deviceHealth{
+						DevID:  monDevID,
+						Health: pluginapi.Unhealthy,
+					}
+				}
+			} else if event.Name == dp.socketPath && event.Op == fsnotify.Remove {
+				logrus.Infof("device socket file for device %s was removed, kubelet probably restarted.", dp.resourceName)
+				return nil
+			}
+		}
+	}
+}
+
+func (dp *VGPUDevicePlugin) GetDeviceName() string {
+	return dp.resourceName
+}
+
+// Stop stops the gRPC server
+func (dp *VGPUDevicePlugin) stopDevicePlugin() error {
+	defer func() {
+		if !IsChanClosed(dp.done) {
+			close(dp.done)
+		}
+	}()
+
+	// Give the device plugin one second to properly deregister
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+	select {
+	case <-dp.deregistered:
+	case <-ticker.C:
+	}
+
+	dp.server.Stop()
+	dp.setInitialized(false)
+	return dp.cleanup()
+}
```

Comment:
I think I found a possible deadlock here, which is hidden by the ticker. Pcidevice controller waits for the chan `dp.deregistered`, but it only be notified if `ListAndWatch()` is [knocked by chan done](https://github.com/ibrokethecloud/pcidevices/blob/3a762072ebdd1506de8db95f5b829ea05f48e23d/pkg/deviceplugins/vgpu_device_manager.go#L197-L198)  and [close chan `dp.deregistered`](https://github.com/ibrokethecloud/pcidevices/blob/3a762072ebdd1506de8db95f5b829ea05f48e23d/pkg/deviceplugins/vgpu_device_manager.go#L209), however, `stopDevicePlugin()` never close chan `dp.done` first.

As a result, `stopDevicePlugin()` relies on ticker timeout to stop the server and remove the socket file, after this `ListAndWatch()` finally gets the chance to [get out of the loop](https://github.com/ibrokethecloud/pcidevices/blob/3a762072ebdd1506de8db95f5b829ea05f48e23d/pkg/deviceplugins/vgpu_device_manager.go#L183-L203) and try to [send empty list response](https://github.com/ibrokethecloud/pcidevices/blob/3a762072ebdd1506de8db95f5b829ea05f48e23d/pkg/deviceplugins/vgpu_device_manager.go#L206-L208), but during this phase, the server already torn down, so it will receive err message `nvidia.com/NVIDIA_xx-xxx device plugin failed to deregister: rpc error: code = Unavailable desc = transport is closing`

Reply:
this should now be fixed with some minor tweaks to logic in `stopDevicePlugin`